client/rest/src/main/java/org/elasticsearch/client/ResponseListener.java

```diff
@@ -19,24 +19,22 @@
 
 package org.elasticsearch.client;
 
-import org.apache.http.StatusLine;
-import org.apache.http.client.methods.CloseableHttpResponse;
-import org.apache.http.message.BasicHttpResponse;
-
-import java.io.IOException;
-
 /**
- * Simple {@link CloseableHttpResponse} impl needed to easily create http responses that are closeable given that
- * org.apache.http.impl.execchain.HttpResponseProxy is not public.
+ * Listener to be provided when calling async performRequest methods provided by {@link RestClient}.
+ * Those methods that do accept a listener will return immediately, execute asynchronously, and notify
+ * the listener whenever the request yielded a response, or failed with an exception.
  */
-class CloseableBasicHttpResponse extends BasicHttpResponse implements CloseableHttpResponse {
+public interface ResponseListener {
 
-    public CloseableBasicHttpResponse(StatusLine statusline) {
-        super(statusline);
-    }
+    /**
+     * Method invoked if the request yielded a successful response
+     */
+    void onSuccess(Response response);
```

Comment:
Indeed, this is the way we allow sniffing on failure internally. For each single failure, regardless of the final outcome of that request, which may be retried on different nodes, we want to mark the host failed, optionally sniff on failure and schedule a new sniff round rather sooner than later. FailureListener is not supposed to be implemented by users, and that is why I don't want to get it mixed the ResponseListener. ResponseListener is the one that counts for users of our async methods. Does this clarify things?